### DISCUSSION TITLE
keep_alive to non-argument and non-return value

### DISCUSSION THREAD

**Discussion by caffeinism** [pybind/pybind11#4139] [Q&A] [answered]
hello! I'm having a hard time getting this to work. What this method does is copies the vector, clears the original vector, and then returns an iterator of the copied vector.

```
.def("iter_and_clear", [](std::vector<something> &self) {
            auto copy = make_shared<std::vector<something>>(self);
            self.clear();
            return py::make_iterator(copy->begin(), copy->end());
        }
```

At the moment of executing this function, contrary to my wish, "copy" is freed and an error occurs as soon as it returns, because std::iterator does not refer to "copy".

I've read the keep_alive documentation, but this seems to only apply to arguments and return values. Is there any way to apply a similar approach to other variables?

**Comment by caffeinism** [comment, marked as answer]
I solved it by returning a vector and defining an __iter__ of the vector.

**Reply by nickguo** [reply]
Is that `iter` a C++ object, or an `iter()` call on the python side?

I am running into the exact same issue and am trying to come up with a C++ only solution if possible.
